### PR TITLE
This winter solstice, we celebrate the wonders of the parallax effect

### PR DESCRIPTION
**Pull request by gvarnavi**
Various updates and new functionality to parallax.  
Notably:

- Realspace mask during preprocess
```python
parallax.preprocess(
    realspace_mask = mask,
    apply_realspace_mask_to_stack=True, # default
)
```
- Explicit bin setting during reconstruct
```python
parallax.reconstruct(
    alignment_bin_values = [32,16,8,8,4,4,2,2,2,2]
)
```
- Improvements to aberration fitting:
  - `force_rotation_deg`
  - explicitly setting aberrations to include in fit
  ```python
  parallax.aberration_fit(
      fit_BF_shifts=True,
      fit_aberrations_mn = [[1,0,0],[2,1,0],[2,1,1]], # defocus and coma only
  )
  ```
  - convert fitted aberrations to polar dictionary for use in ptycho
  ```python
  ptycho = py4DSTEM.process.phase.SingleslicePtychographicReconstruction(
      polar_parameters = parallax.aberration_dict_polar
  )
  ```

- Improvements to upsampling interpolation methods
  - Correct DF and BF warning limits
  - Bilinear interpolation (default)
  - Lanczos interpolation
  ```python
  parallax.subpixel_alignment(
      lanczos_interpolation_order = 2
  )
  ```
  - Nearest-neighbor rolling
  ```python
  parallax.subpixel_alignment(
      integer_pixel_rolling_alignment = True
  )
  ```
- Probe position refinement

Fixed-step gradient descent or "checker-boarding" algorithm with `initial_step_size`, decreasing geometrically by `step_size_factor` up to a minimum step of `min_step_size`.

  ```python
  parallax.subpixel_alignment(
      position_correction_num_iter = 2,
      position_correction_checkerboard_steps = False, # default
      position_correction_initial_step_size=0.125,
      position_correction_min_step_size=0.1,
      position_correction_step_size_factor=0.75,
      position_correction_butterworth_q_lowpass = (0.05,0.05),
      position_correction_butterworth_q_highpass = (0.005,0.005),
      position_correction_butterworth_order = (4,4),
  )
  ```

Probe-position refinement results can be inspected using
```python
parallax.show_probe_position_shifts(
     # vmin=-0.5,
     # vmax=0.5,
)
```

### CODE REVIEW

File: py4DSTEM/process/phase/utils.py [hunk 1]
```diff
@@ -413,7 +413,8 @@ def polar_coordinates(self, x, y):
         """Calculate a polar grid for a given Cartesian grid."""
         xp = self._xp
         alpha = xp.sqrt(x.reshape((-1, 1)) ** 2 + y.reshape((1, -1)) ** 2)
-        phi = xp.arctan2(x.reshape((-1, 1)), y.reshape((1, -1)))
+        # phi = xp.arctan2(x.reshape((-1, 1)), y.reshape((1, -1))) # bug in abtem-legacy and py4DSTEM<=0.14.9
+        phi = xp.arctan2(y.reshape((1, -1)), x.reshape((-1, 1)))
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:417]: This is fine as-is but I would point out that `.ravel()` does the same thing (and won't make a copy in this case)

Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:417]:
```suggestion
        phi = xp.arctan2(y.ravel(), x.ravel())
```

File: py4DSTEM/process/phase/utils.py [hunk 2]
```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
...
+):
+    """
+    Bilinear sampling of intensities from an image array and pixel positions.
+
+    Parameters
+    ----------
+    image: np.ndarray
+        Image array to sample from
+    xa: np.ndarray
+        Vertical positions of image array in pixels
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:1694]:
```suggestion
        Vertical interpolation sampling positions of image array in pixels
```

File: py4DSTEM/process/phase/utils.py [hunk 3]
```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
...
+    ya: np.ndarray
+        Horizontal positions of image array in pixels
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:1696]:
```suggestion
        Horizontal interpolation sampling positions of image array in pixels
```

File: py4DSTEM/process/phase/utils.py [hunk 4]
```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
...
+):
+    """
+    Lanczos sampling of intensities from an image array and pixel positions.
+
+    Parameters
+    ----------
+    image: np.ndarray
+        Image array to sample from
+    xa: np.ndarray
+        Vertical positions of image array in pixels
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:1759]:
```suggestion
        Vertical Interpolation sampling positions of image array in pixels
```

File: py4DSTEM/process/phase/utils.py [hunk 5]
```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
...
+    ya: np.ndarray
+        Horizontal positions of image array in pixels
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:1761]:
```suggestion
        Horizontal interpolation sampling positions of image array in pixels
```

File: py4DSTEM/process/phase/utils.py [hunk 6]
```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
...
+    xF = xp.floor(xa).astype("int")
+    yF = xp.floor(ya).astype("int")
+    dx = xa - xF
+    dy = ya - yF
+
+    all_inds = []
+    all_weights = []
+
+    for i in range(-alpha + 1, alpha + 1):
+        for j in range(-alpha + 1, alpha + 1):
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:1780]: Is this sum meant to be asymmetric about zero? i.e. should it be -alpha to alpha inclusive, or -alpha+1 to alpha as written?

File: py4DSTEM/process/phase/utils.py [hunk 7]
```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
...
+    xp=np,
+    gaussian_filter=gaussian_filter,
+):
+    """
+    kernel density estimate from a set coordinates (xa,ya) and intensity weights.
+
+    Parameters
+    ----------
+    stack: np.ndarray
+        Unshifted virtual BF images stack
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:1822]:
```suggestion
        Unshifted image stack
```

File: py4DSTEM/process/phase/utils.py [hunk 8]
```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
...
+    shifts: np.ndarray
+        Cross-correlated virtual BF image shifts
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:1824]:
```suggestion
        Shifts for each image in stack, shape: (?,?)
```

File: py4DSTEM/process/phase/utils.py [hunk 9]
```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
...
+        scale = np.array(scale)
+        if scale.size == 1:
+            scale = np.tile(scale, 2)
+
+        output_size = (input_size * scale).astype("int")
+    else:
+        if output_size is None:
+            raise ValueError()
+        if output_size.size != 2:
+            raise ValueError()
```
Comment (sezelt) [inline py4DSTEM/process/phase/utils.py:2101]: Needs docstring and error messages

File: py4DSTEM/process/phase/iterative_parallax.py [hunk 10]
```diff
@@ -251,23 +260,26 @@ def _populate_instance(self, group):
 
     def preprocess(
         self,
-        edge_blend: int = 16,
+        edge_blend: float = 16.0,
+        dp_mask: np.ndarray = None,
```
Comment (sezelt) [inline py4DSTEM/process/phase/iterative_parallax.py:264]: Needs entry in docstring

**Review by sezelt — APPROVED** [review]
All looks good, mostly just highlighted a few missing/incomplete docstring things

File: py4DSTEM/process/phase/utils.py (hunk 1 above)
Comment (gvarnavi) [reply py4DSTEM/process/phase/utils.py:417]: I thought `.ravel()` has the same copy-behaviour as `.reshape()`?  
This was originally copied from abtem (hence the origin of the pesky bug), but can however be simply broadcasted:
```python
alpha = xp.sqrt(x[:,None]**2 + y[None,:]**2)
phi = xp.arctan2(y[None,:], x[:,None])
```

Comment (gvarnavi) [reply py4DSTEM/process/phase/utils.py:417]: This needs to be broadcasted, see above.

File: py4DSTEM/process/phase/utils.py (hunk 6 above)
Comment (gvarnavi) [reply py4DSTEM/process/phase/utils.py:1780]: This is correct as-written, the interpolation weights are computed on a [0,1] grid (same as we do for bilinear). See equation here: https://pixinsight.com/doc/docs/InterpolationAlgorithms/InterpolationAlgorithms.html

**Pull request merged**